src/lib.rs

```diff
@@ -109,8 +111,22 @@ mod alias;
 #[cfg(feature="nightly")]
 pub use alias::*;
 
+use std::ops::{Deref, DerefMut};
+
 /// Abstract collection.
 pub trait Collection {
+	/// Type of references to items of the collection.
+	type ItemRef<'a>: Deref<Target = Self::Item>
```

Comment:
I there a reason why you choose to add the `Self: 'a` where clause? Why not just
```rust
type ItemRef<'a>: 'a + Deref<Target = Self::Item>
```

Reply:
> I there a reason why you choose to add the `Self: 'a` where clause? Why not just
> 
> ```rust
> type ItemRef<'a>: 'a + Deref<Target = Self::Item>
> ```

I chose to as `Self: 'a` because any generics placed in the type must live as long as `'a`. `'a` in this context is unbounded, so it could be anything, which would require that generics placed in the type must live long as `'static`.

In theory, the bound `Self::Item: 'a` would work fine, but in practice, some reference types may hold more references to generic types than just `Self::Item`, e.g. [`DashMap`'s reference type](https://docs.rs/dashmap/4.0.2/dashmap/mapref/one/struct.Ref.html) also holds a reference to the key.

The most correct solution would probably be to add another type for every reference, where implementors could provide all the generics their reference type uses, so we could add a bound `Self::ItemRefGenerics: 'a` or something like that. But just adding `Self: 'a` probably works just as well and is more elegant.

Reply:
I'm not sure to see what you mean. `dashmap::mapref::one::Ref<'a, K, V, S>` has only one lifetime parameter here. If we take the `Get` trait for instance with the `get` function:
```rust
fn get<'a>(&'a self, key: T) -> Option<Self::ItemRef<'a>>;
```
The `'a` lifetime on `self` should imply `Self: 'a`. Moreover I think associated type lifetimes are implicitly bound by the implementor's lifetime right? Here is an example:
```rust
#![feature(generic_associated_types)]
pub trait Foo {
  type T<'a>;
}

impl<'t, T> Foo for &'t T {
  type T<'a> = Self;
}
```
This compiles even though I never explicitly specified `Self: 'a` for `Foo:T<'a>`. But it automatically derived `&'t T: 'a` even though `&'t T` is obviously not `'static`.

Reply:
In fact, following this reasoning I even wonder if specifying `Deref<Target = Self::Item>` is enough, instead of `'a + Deref<Target = Self::Item>`...

Reply:
Hu... you are right, the `Self: 'a` bound seems required where the implementor has type parameters. The implicit bound that seems to be automatically derived for `Self` in my example is not automatically derived for the type parameters. How weird.

Reply:
No it is even weirder. In my example It doesn't work if we replace `Self` with `&'a Self`. So it can detect `Self: 'a` but not `&'a Self: 'a`? I feel like my example should not be accepted in the first place...

Reply:
My confusion has been resolved [in the forum](https://users.rust-lang.org/t/generic-associated-types-lifetime/63671). We will need to wait for [#87479](https://github.com/rust-lang/rust/issues/87479) to make an actual decision here about the `where Self: 'a` bound.